src/main/java/io/temporal/internal/statemachines/TimerStateMachine.java

```diff
@@ -0,0 +1,179 @@
+                  State.START_COMMAND_RECORDED,
+                  EntityStateMachineInitialCommand::setInitialCommandEventId)
+              .add(
+                  State.START_COMMAND_CREATED,
+                  ExplicitEvent.CANCEL,
+                  State.CANCELED,
+                  TimerStateMachine::cancelStartTimerCommand)
+              .add(
+                  State.START_COMMAND_RECORDED,
+                  EventType.EVENT_TYPE_TIMER_FIRED,
+                  State.FIRED,
+                  TimerStateMachine::notifyCompletion)
+              .add(
+                  State.START_COMMAND_RECORDED,
+                  ExplicitEvent.CANCEL,
+                  State.CANCEL_TIMER_COMMAND_CREATED,
+                  TimerStateMachine::createCancelTimerCommand)
+              .add(
+                  State.CANCEL_TIMER_COMMAND_CREATED,
+                  ExplicitEvent.CANCEL,
+                  State.CANCEL_TIMER_COMMAND_CREATED)
+              .add(
+                  State.CANCEL_TIMER_COMMAND_CREATED,
+                  EventType.EVENT_TYPE_TIMER_CANCELED,
+                  State.CANCELED)
+              .add(
+                  State.CANCEL_TIMER_COMMAND_CREATED,
+                  CommandType.COMMAND_TYPE_CANCEL_TIMER,
+                  State.CANCEL_TIMER_COMMAND_CREATED,
+                  TimerStateMachine::notifyCancellation)
+              .add(
+                  State.CANCEL_TIMER_COMMAND_CREATED,
+                  EventType.EVENT_TYPE_TIMER_FIRED,
+                  State.FIRED,
+                  TimerStateMachine::cancelTimerCommandFireTimer);
+
+  private void createStartTimerCommand() {
+    addCommand(
+        Command.newBuilder()
+            .setCommandType(CommandType.COMMAND_TYPE_START_TIMER)
+            .setStartTimerCommandAttributes(startAttributes)
+            .build());
+  }
+
+  public void cancel() {
+    if (!isFinalState()) {
+      explicitEvent(ExplicitEvent.CANCEL);
+    }
+  }
+
+  private void cancelStartTimerCommand() {
+    cancelCommand();
+    notifyCancellation();
+  }
+
+  private void notifyCancellation() {
+    completionCallback.apply(
+        HistoryEvent.newBuilder()
+            .setEventType(EventType.EVENT_TYPE_TIMER_CANCELED)
+            .setTimerCanceledEventAttributes(
+                TimerCanceledEventAttributes.newBuilder()
+                    .setIdentity("workflow")
+                    .setTimerId(startAttributes.getTimerId()))
+            .build());
+  }
+
+  private void notifyCompletion() {
+    completionCallback.apply(currentEvent);
+  }
+
+  private void createCancelTimerCommand() {
+    addCommand(
+        Command.newBuilder()
+            .setCommandType(CommandType.COMMAND_TYPE_CANCEL_TIMER)
+            .setCancelTimerCommandAttributes(
+                CancelTimerCommandAttributes.newBuilder().setTimerId(startAttributes.getTimerId()))
+            .build());
+  }
+
+  private void cancelTimerCommandFireTimer() {
+    cancelCommand();
```

Comment:
just curious: what would happen if we did not call cancelCommand() here?

Reply:
Right, it will be sent to the service.